ArdusatSDK.cpp

```diff
@@ -453,266 +217,1101 @@ const char * valuesToCSV(const char *sensorName, unsigned long timestamp, int nu
+ * @return  sensor readings in JSON format or empty string if uninitialized
+ */
+const char * Sensor::readToJSON(const char * sensorName) {
+  this->read();
+  return this->toJSON(sensorName);
+}
 
-  dtostrf(input.lux, 2, 3, _getOutBuf() + _output_buf_len);
-  _output_buf_len = strlen(_getOutBuf());
-  _add_checksum_to_csv_buffer(sensorName, 1, input.lux);
-  _getOutBuf()[_output_buf_len++] = '\n';
 
-  return _getOutBuf();
+/**************************************************************************//**
+ * @brief   Constructs Acceleration sensor object
+ * @ingroup acceleration
+ *
+ * Example Usage:
+ * @code
+ *     Acceleration accel;                        // Instantiate sensor object
+ *     accel.begin();                             // Initialize sensor
+ *     Serial.println(accel.readToJSON("accel")); // Read and print values in JSON
+ * @endcode
+ *****************************************************************************/
+Acceleration::Acceleration(void) :
+  gGain(LSM303_ACCEL_GAIN8G)
+{
+  this->initializeHeader(SENSORID_ADAFRUIT9DOFIMU, DATA_UNIT_METER_PER_SECONDSQUARED, acceleration_sensor_name);
 }
 
-_DEFAULT_SENSOR_NAME_TO_CSV_FXN(uvlight)
-const char * uvlightToCSV(const char *sensorName, uvlight_t & input) {
-  _headerToCSV(&(input.header), sensorName);// warning : resets the internal buffer
+/**
+ * @brief   Sets the gain configuration variable
+ * @ingroup acceleration
+ *
+ * @param gGain Advanced configuration for accelerometer's gain
+ *     - LSM303_ACCEL_GAIN2G
+ *     - LSM303_ACCEL_GAIN4G
+ *     - LSM303_ACCEL_GAIN6G
+ *     - LSM303_ACCEL_GAIN8G (Default)
+ *     - LSM303_ACCEL_GAIN16G
+ */
+void Acceleration::configure(lsm303_accel_gain_e gain) {
+  this->gGain = gain;
+  this->initialized = false;
+}
 
-  dtostrf(input.uvindex, 2, 3, _getOutBuf() + _output_buf_len);
-  _output_buf_len = strlen(_getOutBuf());
-  _add_checksum_to_csv_buffer(sensorName, 1, input.uvindex);
-  _getOutBuf()[_output_buf_len++] = '\n';
+/**
+ * @brief   Initializes the sensor with any set configurations or defaults
+ * @ingroup acceleration
+ *
+ * @retval true  Successfully initialized
+ * @retval false Failed to initialize
+ */
+boolean Acceleration::initialize(void) {
+  return lsm303_accel_init(this->gGain);
+}
 
-  return _getOutBuf();
+/**
+ * @brief   Takes a reading from the sensor
+ * @ingroup acceleration
+ */
+void Acceleration::read(void) {
+  this->header.timestamp = millis();
+  lsm303_getAccel(&(this->x), &(this->y), &(this->z));
 }
 
-const char * orientationToCSV(const char *sensorName, orientation_t & input) {
-  _headerToCSV(&(input.header), sensorName);
+/**
+ * @brief   Returns last read value in CSV format
+ * @ingroup acceleration
+ * @param   sensorName The text to display next to the value
+ * @return  sensor readings in CSV format or empty string if uninitialized
+ */
+const char * Acceleration::toCSV(const char * sensorName) {
+  if (this->initialized) {
```

Comment:
I'm not sure this is really the best check for this -> For purposes of `toCSV`, what we're really checking isn't weather the sensor is "initialized" (e.g. hardware check, begin & any hardware configuration has been performed), but whether we've had a successful read (which in turn implies that the sensor is or at least once was initializes). Maybe set the `timestamp` in the `header` to an invalid/unread initial value (e.g. -1 if it's signed, 0 if unsigned) that we can check here instead?